src/server/data_source.rs

```diff
@@ -1,142 +1,197 @@
-            let node_context = unsafe { NodeContext::peek_at(node_context) };
-            #[allow(irrefutable_let_patterns)] // We will add more node context types eventually.
-            let NodeContext::DataSource(data_source) = node_context
-            else {
-                // We expect to always find this node context type.
-                return ua::StatusCode::BADINTERNALERROR.into_raw();
-            };
-            let DataSource { read, .. } = data_source;
-
-            // TODO: Prepare callback context.
-            let mut context = DataSourceReadContext;
-
-            match read(&mut context) {
-                Ok(()) => ua::StatusCode::GOOD,
-                Err(status_code) => status_code,
-            }
-            .into_raw()
-        }
+    /// [`set_value()`]: Self::set_value
+    pub fn set_variant(&mut self, variant: ua::Variant) {
+        self.set_value(ua::DataValue::new(variant));
+    }
+}
 
-        unsafe extern "C" fn write_c(
-            _server: *mut UA_Server,
-            _session_id: *const UA_NodeId,
-            _session_context: *mut c_void,
-            _node_id: *const UA_NodeId,
-            node_context: *mut c_void,
-            _range: *const UA_NumericRange,
-            _value: *const UA_DataValue,
-        ) -> UA_StatusCode {
-            let node_context = unsafe { NodeContext::peek_at(node_context) };
-            #[allow(irrefutable_let_patterns)] // We will add more node context types eventually.
-            let NodeContext::DataSource(data_source) = node_context
-            else {
-                // We expect to always find this node context type.
-                return ua::StatusCode::BADINTERNALERROR.into_raw();
-            };
-            let DataSource { write, .. } = data_source;
-            let Some(write) = write else {
-                return ua::StatusCode::BADWRITENOTSUPPORTED.into_raw();
-            };
-
-            // TODO: Prepare callback context.
-            let mut context = DataSourceWriteContext;
-
-            match write(&mut context) {
-                Ok(()) => ua::StatusCode::GOOD,
-                Err(status_code) => status_code,
+/// Context when [`DataSource`] is being written to.
+#[allow(clippy::module_name_repetitions)]
+pub struct DataSourceWriteContext {
+    /// Incoming value to be written.
+    ///
+    /// This is a non-mutable (const) cell where the write callback receives the data to be written
+    /// by the client.
+    value_source: NonNull<UA_DataValue>,
+}
+
+impl DataSourceWriteContext {
+    /// Creates context for write callback.
+    fn new(value: *const UA_DataValue) -> Option<Self> {
+        Some(Self {
+            // SAFETY: `NonNull`
+            value_source: NonNull::new(value.cast_mut())?,
+        })
+    }
+
+    /// Gets value.
+    ///
+    /// This returns the value received from the client that is writing to this [`DataSource`].
+    #[must_use]
+    pub fn value(&self) -> &ua::DataValue {
```

Comment:
Would it be possible to obtain ownership and consume the value exactly once? Or doesn't this make sense?